[TEST-FAILURE] ListenerTest.replaceFiresUpdatedEvent

```
java.lang.AssertionError: expected:<0> but was:<1>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:743)
    at org.junit.Assert.assertEquals(Assert.java:118)
    at org.junit.Assert.assertEquals(Assert.java:555)
    at org.junit.Assert.assertEquals(Assert.java:542)
    at com.hazelcast.map.ListenerTest$5.run(ListenerTest.java:246)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:236)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:252)
    at com.hazelcast.map.ListenerTest.replaceFiresUpdatedEvent(ListenerTest.java:243)
``
`
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK8/com.hazelcast$hazelcast/310/testReport/junit/com.hazelcast.map/ListenerTest/replaceFiresUpdatedEvent/
see also: #3156
```

--- a/hazelcast/src/main/java/com/hazelcast/map/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/DefaultRecordStore.java
@@ -775,47 +775,47 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore implements 
 
     // TODO why does not replace method load data from map store if currently not available in memory.
     @Override
-    public Object replace(Data key, Object value) {
+    public Object replace(Data key, Object update) {
         checkIfLoaded();
-        final long now = getNow();
 
-        Record record = getRecordOrNull(key, false);
-        Object oldValue;
-        if (record != null && record.getValue() != null) {
-            oldValue = record.getValue();
-            value = mapServiceContext.interceptPut(name, oldValue, value);
-            value = mapDataStore.add(key, value, now);
-            onStore(record);
-            updateSizeEstimator(-calculateRecordHeapCost(record));
-            setRecordValue(record, value, now);
-            updateSizeEstimator(calculateRecordHeapCost(record));
-        } else {
+        final Record record = getRecordOrNull(key, false);
+        if (record == null || record.getValue() == null) {
             return null;
         }
+        final Object oldValue = record.getValue();
+        update = mapServiceContext.interceptPut(name, oldValue, update);
+        final long now = getNow();
+        update = mapDataStore.add(key, update, now);
+        onStore(record);
+        updateSizeEstimator(-calculateRecordHeapCost(record));
+        setRecordValue(record, update, now);
+        updateSizeEstimator(calculateRecordHeapCost(record));
         saveIndex(record);
         evictEntries(now, false);
         return oldValue;
     }
 
     @Override
-    public boolean replace(Data key, Object testValue, Object newValue) {
+    public boolean replace(Data key, Object expect, Object update) {
         checkIfLoaded();
-        final long now = getNow();
 
-        Record record = getRecordOrNull(key, false);
+        final Record record = getRecordOrNull(key, false);
         if (record == null) {
             return false;
         }
-        if (mapServiceContext.compare(name, record.getValue(), testValue)) {
-            newValue = mapServiceContext.interceptPut(name, record.getValue(), newValue);
-            newValue = mapDataStore.add(key, newValue, now);
-            onStore(record);
-            updateSizeEstimator(-calculateRecordHeapCost(record));
-            setRecordValue(record, newValue, now);
-            updateSizeEstimator(calculateRecordHeapCost(record));
-        } else {
+        final MapServiceContext mapServiceContext = this.mapServiceContext;
+        final Object current = record.getValue();
+        final String mapName = this.name;
+        if (!mapServiceContext.compare(mapName, current, expect)) {
             return false;
         }
+        update = mapServiceContext.interceptPut(mapName, current, update);
+        final long now = getNow();
+        update = mapDataStore.add(key, update, now);
+        onStore(record);
+        updateSizeEstimator(-calculateRecordHeapCost(record));
+        setRecordValue(record, update, now);
+        updateSizeEstimator(calculateRecordHeapCost(record));
         saveIndex(record);
         evictEntries(now, false);
         return true;

--- a/hazelcast/src/main/java/com/hazelcast/map/RecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/RecordStore.java
@@ -74,9 +74,21 @@ public interface RecordStore {
 
     boolean containsKey(Data dataKey);
 
-    Object replace(Data dataKey, Object value);
+    Object replace(Data dataKey, Object update);
 
-    boolean replace(Data dataKey, Object oldValue, Object newValue);
+
+    /**
+     * Sets the value to the given updated value
+     * if {@link com.hazelcast.map.record.RecordFactory#isEquals} comparison
+     * of current value and expected value is {@code true}.
+     *
+     * @param dataKey key which's value is requested to be replaced.
+     * @param expect  the expected value
+     * @param update  the new value
+     * @return {@code true} if successful. False return indicates that
+     * the actual value was not equal to the expected value.
+     */
+    boolean replace(Data dataKey, Object expect, Object update);
 
     void putTransient(Data dataKey, Object value, long ttl);
 

--- a/hazelcast/src/main/java/com/hazelcast/map/operation/ReplaceIfSameOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/operation/ReplaceIfSameOperation.java
@@ -20,25 +20,26 @@ import com.hazelcast.nio.IOUtil;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.Data;
+
 import java.io.IOException;
 
 public class ReplaceIfSameOperation extends BasePutOperation {
 
-    private Data testValue;
+    private Data expect;
     private boolean successful;
 
-    public ReplaceIfSameOperation(String name, Data dataKey, Data testValue, Data value) {
-        super(name, dataKey, value);
-        this.testValue = testValue;
+    public ReplaceIfSameOperation(String name, Data dataKey, Data expect, Data update) {
+        super(name, dataKey, update);
+        this.expect = expect;
     }
 
     public ReplaceIfSameOperation() {
     }
 
     public void run() {
-        successful = recordStore.replace(dataKey, testValue, dataValue);
+        successful = recordStore.replace(dataKey, expect, dataValue);
         if (successful) {
-            dataOldValue = testValue;
+            dataOldValue = expect;
         }
     }
 
@@ -64,13 +65,13 @@ public class ReplaceIfSameOperation extends BasePutOperation {
     @Override
     protected void writeInternal(ObjectDataOutput out) throws IOException {
         super.writeInternal(out);
-        IOUtil.writeNullableData(out, testValue);
+        IOUtil.writeNullableData(out, expect);
     }
 
     @Override
     protected void readInternal(ObjectDataInput in) throws IOException {
         super.readInternal(in);
-        testValue = IOUtil.readNullableData(in);
+        expect = IOUtil.readNullableData(in);
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/map/proxy/MapProxySupport.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/proxy/MapProxySupport.java
@@ -449,8 +449,8 @@ abstract class MapProxySupport extends AbstractDistributedObject<MapService> imp
         }
     }
 
-    protected boolean replaceInternal(final Data key, final Data oldValue, final Data newValue) {
-        ReplaceIfSameOperation operation = new ReplaceIfSameOperation(name, key, oldValue, newValue);
+    protected boolean replaceInternal(final Data key, final Data expect, final Data update) {
+        ReplaceIfSameOperation operation = new ReplaceIfSameOperation(name, key, expect, update);
         boolean replaceSuccessful = (Boolean) invokeOperation(key, operation);
         invalidateNearCache(key);
         return replaceSuccessful;

--- a/hazelcast/src/test/java/com/hazelcast/map/ListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/ListenerTest.java
@@ -31,15 +31,16 @@ import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ProblematicTest;
 import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
 import java.util.Map;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
-import org.junit.Before;
-import org.junit.Test;
-import org.junit.experimental.categories.Category;
-import org.junit.runner.RunWith;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
@@ -221,30 +222,25 @@ public class ListenerTest extends HazelcastTestSupport {
      */
     @Test
     public void replaceFiresUpdatedEvent() throws InterruptedException {
-        TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(1);
-        HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
-        IMap<Object, Object> map = h1.getMap(randomMapName());
-        map.put(1, 1);
-        final AtomicInteger updateCount = new AtomicInteger(0);
-        final AtomicInteger addCount = new AtomicInteger(0);
-        map.addEntryListener(new EntryAdapter<Object, Object>() {
-            @Override
-            public void entryAdded(EntryEvent<Object, Object> event) {
-                addCount.incrementAndGet();
-            }
+        final AtomicInteger entryUpdatedEventCount = new AtomicInteger(0);
 
+        HazelcastInstance node = createHazelcastInstance();
+        IMap<Integer, Integer> map = node.getMap(randomMapName());
+        map.put(1, 1);
+
+        map.addEntryListener(new EntryAdapter<Integer, Integer>() {
             @Override
-            public void entryUpdated(EntryEvent<Object, Object> event) {
-                updateCount.incrementAndGet();
+            public void entryUpdated(EntryEvent<Integer, Integer> event) {
+                entryUpdatedEventCount.incrementAndGet();
             }
         }, true);
-        map.replace(1, 1, 2);  // should succeed and fire entryUpdated() exactly once
-        map.replace(1, 1, 2);  // should fail
+
+        map.replace(1, 1, 2);
+
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() {
-                assertEquals(0, addCount.get());
-                assertEquals(1, updateCount.get());
+                assertEquals(1, entryUpdatedEventCount.get());
             }
         });
     }

